lib/command.js

```diff
@@ -230,14 +230,19 @@ module.exports = function command (yargs, usage, validation, globalMiddleware) {
 
       innerArgv = applyMiddleware(innerArgv, yargs, middlewares, false)
 
-      const handlerResult = isPromise(innerArgv)
-        ? innerArgv.then(argv => commandHandler.handler(argv))
-        : commandHandler.handler(innerArgv)
+      let handlerResult
+      if (isPromise(innerArgv)) {
+        yargs.getUsageInstance().cacheHelpMessage()
+        handlerResult = innerArgv.then(argv => commandHandler.handler(argv))
+      } else {
+        handlerResult = commandHandler.handler(innerArgv)
+      }
 
       if (isPromise(handlerResult)) {
-        handlerResult.catch(error =>
+        yargs.getUsageInstance().cacheHelpMessage()
+        handlerResult.catch(error => {
           yargs.getUsageInstance().fail(null, error)
```

Comment:
Your unhandled rejection comes from here: when exitProcess is set to false, the last thing fail does is throwing, so your catch handler throws itself, the promise is still rejected, and this rejection is not handled.
```suggestion
          try {
            yargs.getUsageInstance().fail(null, error)
          }
          catch() {
          }
```